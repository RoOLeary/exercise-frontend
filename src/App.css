.app {
  text-align: center;
}

.show-component {
  text-align: center;
  padding: 0;
  margin-top: 10px;
  user-select: none;
}

.show-image {
  width: 250px;
  height: 375px;
}

.show-episodes {
  color: #aaa;
  font-family: HelveticaNeueLT, "Helvetica Neue", Helvetica, Arial, sans-serif;
  font-weight: bold;
  font-size: small;
  padding: 0;
  margin: 10px 0 0 5px;
}

.show-title {
  font-weight: bold;
  text-transform: uppercase;
  font-family: HelveticaNeueLT, "Helvetica Neue", Helvetica, Arial, sans-serif;
  padding: 0;
  margin: 5px 0 10px 5px;
}

.show-navigation {
  font-size: 30px;
  padding: 10px 40px 10px 40px;
  margin: 10px 0 0 10px;
}

@media (min-width: 980px) {

  .show-image {
    width: 360px;
    height: 540px;
  }

  .show-navigation {
    font-size: 20px;
    padding: 10px 20px 10px 20px;
    margin: 10px 0 0 10px;
  }

}
