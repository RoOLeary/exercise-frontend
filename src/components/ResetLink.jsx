import React from 'react';
import styled from 'styled-components';

const Reset = styled.a`
    padding: 0.5em;
    font-family: 'Space Grotesk',sans-serif;
    background-color: #f5f5f5;
    border-radius: 5px;
    transition: all .5s ease;
    font-size: .75em;

    :hover{
        background-color: #000;
        color: #fff;
    }
`;

const resetComponent = () => {
    sessionStorage.removeItem("saveCurrentActive");
    sessionStorage.removeItem("saveCurrentIndex");
    window.location.reload();
}


const ResetLink = () => {
    return(
        <Reset onClick={() => resetComponent()}>Reset</Reset>
    )
}

export default ResetLink;