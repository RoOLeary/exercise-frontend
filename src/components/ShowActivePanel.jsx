import React from 'react';
import { useContext } from 'react';
import { DataContext } from '../contexts/dataContext';
import styled from 'styled-components';


const Title = styled.h1`
    font-size: 1em;
    font-family: 'Montserrat', sans-serif;
    font-weight: bold;
    text-transform: uppercase;
    letter-spacing: 2px;
    line-height: 1.25em;
`;

const Episodes = styled.p`
    font-size: .75em;
    font-family: 'Space Grotesk', sans-serif;
    text-transform: uppercase;
    letter-spacing: .75px;
    color: darkgray;
`;

const MainStage = styled.div`
    width: 100%;
    margin: 0 auto;
    text-align: center;
    transition: all .5s ease;
    @media screen and (max-width: 980px){
        margin-top: 2em;
    }
   
`;

const Image = styled.img`
    width: 250px;
    @media screen and (max-width: 980px){
        margin-top: 2em;
    }
`;

const ShowActivePanel = () => {

    const dataContext = useContext(DataContext);
    const { shows, isActive } = dataContext.state;
    const show = shows ? shows.filter(show => show.id === isActive) : '';
    
    const Output = Object.entries(show).map((s, i) => {
        const { title, episodes, product_image_url } = s[1];
        return ( 
            <MainStage className={`shows-main`}  key={i}>
                <Image src={`${product_image_url}`} alt={title} />
                <Episodes>{episodes} episodes</Episodes>
                <Title>{title}</Title> 
            </MainStage>
        )
    });

    return Output
}

export default ShowActivePanel;