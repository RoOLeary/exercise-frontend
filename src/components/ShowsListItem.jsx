import React from 'react';
import { useContext } from 'react';
import { DataContext } from '../contexts/dataContext';
import styled from 'styled-components';


const Box = styled.div`
    width: 50px;
    height: 50px;
    background-color: ${props => props.activeSlide ? 'black' : 'lightgray' };
    margin-bottom: 10px;
    transition: all .5s ease;
    &:hover{
        background-color: ${props => props.activeSlide ? 'black' : 'darkgray' };
    }
`;

const Indicator = styled.span`
    display: ${props => props.activeSlide ? 'block' : 'none' };
    font-family: 'Space Grotesk', sans-serif;
`


const ShowPanelListItem = styled.li`
    list-style: none;
    padding: 1em;
    margin-right: 2em;
    text-align: center;
    scroll-snap-align: start;  
    display: inline-block;
    overscroll-behavior: contain;

    @media screen and (max-width: 980px){
        margin-right: 0;
    }
`

const ShowPanel = ( props ) => {
    const dataContext = useContext(DataContext);
    const { index, isActive } = dataContext.state;
    const { setIsActive, setIndex } = dataContext.handlers;
    const id = props.content;
    const activeSlide = (id === isActive) ? 'active' : '';
    // or this....
    var idx = props.index;

    const triggers = (id, idx) => {
        // console.log('set id to: ' + id)
        setIndex({type: "setIndex", payload: idx }); 
        setIndex({type: "nextIndex", payload: ++idx  }); 
        setIndex({type: "prevtIndex", payload: --idx }); 
        setIsActive(id);
    }

    return(
        <ShowPanelListItem id={id} index={idx} className={activeSlide}>
            <>
                <Box activeSlide={activeSlide} onClick={() => triggers(id, idx)} />
                <Indicator activeSlide={activeSlide}>{idx ? ++idx : 1}</Indicator>
            </>
        </ShowPanelListItem>
    )
}

export default ShowPanel;