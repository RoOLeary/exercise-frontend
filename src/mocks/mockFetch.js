const showsListResponse = [{
    "id": "a1",
    "title": "Gaycation",
    "episodes": 24,
    "product_image_url": "/images/gaycation.jpg"
}];

const dogImagesResponse = {
    message: [
        "https://images.dog.ceo/breeds/cattledog-australian/IMG_1042.jpg ",
        "https://images.dog.ceo/breeds/cattledog-australian/IMG_5177.jpg",
    ],
};

export default async function mockFetch(url) {
    switch (url) {
        case "https://localhost:3000/shows.json": {
            return {
                ok: true,
                status: 200,
                json: async () => showsListResponse,
            };
        }
        // case "https://dog.ceo/api/breed/husky/images" :
        // case "https://dog.ceo/api/breed/cattledog/images": {
        //     return {
        //         ok: true,
        //         status: 200,
        //         json: async () => dogImagesResponse,
        //     };
        // }
        default: {
            throw new Error(`Unhandled request: ${url}`);
        }
    }
}