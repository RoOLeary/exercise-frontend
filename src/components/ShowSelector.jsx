import React from 'react';
import { useEffect, useContext, useRef, useCallback } from 'react';
import { DataContext } from '../contexts/dataContext';
import Layout from './Layout';
import ShowsListItem from './ShowsListItem';
import ShowActivePanel from './ShowActivePanel';
import ResetLink from './ResetLink';
import ShowsArrows from './ShowsArrows';
import ToggleArrows from './ToggleArrows';
import styled from 'styled-components';
import './../globals.scss';

/* Styled components */

const Container = styled.div`
    background-color: #ffffff;
    justify-content: center;
    align-items: center;
    margin: 0 auto;
    height: 100vh; 
    display: flex;
    flex-direction: column;
    border: 1px solid rgba(167,170,178,.5);
   
    @media screen and (max-width: 980px){
        flex-direction: column-reverse;
        justify-content: center;

    }
`;


const ShowSlideMenu = styled.ul`
    @media screen and (max-width: 980px){
        width: 280px;
        margin: 0 auto;
    }
    overflow-x: scroll;
    overflow-y: hidden;
    white-space: nowrap;
    width: 500px;
    margin: 0 auto;
    display: grid;
    grid-auto-flow: column;
    overscroll-behavior-x: contain;
    scroll-snap-type: x mandatory;
    position: relative;
    
`;


const Loading = styled.h1`
    text-align: center;
    font-family: 'Montserrat',sans-serif;
    font-size: 16px;
    letter-spacing: 5px;
    text-transform: uppercase;
`;

const Settings = styled.div`
       
`;

const HR = styled.hr`
    width: 80%;
    margin: 2em auto;
    border-top: 1px solid rgba(167,170,178,.5);
    @media screen and (max-width: 980px){   
        width: 100%;
    }
   
`;



const ShowSelector = () => {
    
    // Using DataContext to hydrate the ShowSlide Component. The ShowActivePanel renders the relevant
    // content based on the active show as set by the shared context. 
    
    // By updating the context, am able to manipulate active states etc...without something like Redux. 
    // There are definitely some processing gotcha's here, but in this case seems like it will do the trick.

    const dataContext = useContext(DataContext);
    const { loading, shows, index, isActive } = dataContext.state;


    // so this callback runs on load to check if the active menu element is offscreen, and if so, 
    // to slide on over to the active element. Have duplicated similar elsewhere, which makes me think a further
    // itereation could use this as a custom hook or other helper.  
   
    const checkForValue = useCallback(async(isActive) => {
        let slideToActive = document.querySelector(`#${isActive}`);
        if(slideToActive){
            slideToActive.scrollIntoView({ block: "center", behavior: "smooth"  });
        } 
    }, []); 
    
    // ahem...yeah, so 3/4 second delay to make sure the context has a chance to hydrate. Isn't ideal, but for
    // argument's sake, we can take this as spoofing a server repsonse.  

    useEffect(() => {
        const timeoutId = setTimeout(() => {
            checkForValue(isActive);
        }, 750);

        return () => {
            clearTimeout(timeoutId);
        }
    },[isActive]);
        
    return(
        <Layout>
            <Container>
                <ShowSlideMenu>  
                    {loading ? <Loading>Loading Shows...</Loading> : shows.map((show, i) => {
                        return <ShowsListItem key={i} content={show.id} index={i} />;
                    })}
                </ShowSlideMenu>
                <HR />
                {/* Uncomment the next line and the ToggleArrows component. */}
                {index.show ? <ShowsArrows /> : null}
                <ShowActivePanel />
                
                <Settings>
                    <ResetLink />
                    <ToggleArrows />
                </Settings> 
            </Container>
        </Layout>
    );
};

export default ShowSelector;