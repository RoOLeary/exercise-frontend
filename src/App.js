import React from 'react';
import { Router, Route, IndexRedirect, browserHistory } from 'react-router';
import ShowSelector from './ShowSelector';
const shows = require('./shows.json');
const firstShowId = shows[0].id;

const changeShow = (id) => browserHistory.push(`/shows/${id}`);

const App = ({ params }) => {
  const showId = +params.showId;
  return <div>
    <ShowSelector shows={shows} showId={showId}
                  onChange={changeShow}/>
  </div>;
};

export default () => (
  <Router history={browserHistory}>
    <Route path="/" component={App}>
      <IndexRedirect to={'/shows/' + firstShowId} />
      <Route path="/shows" component={ShowSelector} >
        <IndexRedirect to={'/shows/' + firstShowId} />
      </Route>
      <Route path="/shows/:showId" component={ShowSelector} />
    </Route>
  </Router>
);

