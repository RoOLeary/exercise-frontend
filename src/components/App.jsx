import React from "react";
import ShowSelector from "./ShowSelector";
import DataContextProvider from './../contexts/dataContext';

const App = () => {
    return(
        <DataContextProvider>
            <ShowSelector />
        </DataContextProvider>
    )
}

export default App; 