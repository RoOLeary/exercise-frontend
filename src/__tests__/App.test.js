import React from "react";
import { shallow, render, mount, screen, history } from 'enzyme';
import renderer from 'react-test-renderer';
import App from './../components/App';
import ShowSelector from './../components/ShowSelector';
import ShowsListItem from './../components/ShowsListItem';
import ShowActivePanel from './../components/ShowActivePanel';
import DataContextProvider from './../contexts/dataContext';
import ResetLink from "./../components/ResetLink";
import ToggleArrowsLink from "./../components/ToggleArrows";
import mockFetch from "./../mocks/mockFetch";
import { createMemoryHistory } from 'history'

const jsdom =  require('jsdom');
const { JSDOM } = jsdom;
const { window } = new JSDOM('');
const { document } = (new JSDOM(``)).window;

global.document = document;
global.window = window;

global.fetch = jest.fn(() => Promise.resolve({
    json: () => Promise.resolve(MOCK_SHOW)
}));   

it('renders correctly', () => {
    const tree = renderer
        .create(<DataContextProvider><ShowSelector /></DataContextProvider>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
 
const historyMock = { push: jest.fn() }   
const initialPost = "/?id=a1";

describe('App component', () => {
    

    test('App: render top level container', () => {
      render(<App />);
    });

    test('ShowsListItem: render ShowsListItem', () => { 
        const MOCK_SHOW = {
            "id": "a1",
            "title": "Gaycation",
            "episodes": 24,
            "product_image_url": "/images/gaycation.jpg"
        };
        const tree = shallow(<ShowsListItem content={MOCK_SHOW} />);
        // expect(tree.props().content).toEqual(MOCK_SHOW);
        expect.arrayContaining(MOCK_SHOW);
    });


    test('should render reset button', () => {
        const mockCallBack = jest.fn();
        const reset = shallow(
            <ResetLink onClick={mockCallBack}>
                 Reset
            </ResetLink>
       );
       
       reset.simulate('click');
    //    expect(mockCallBack.mock.calls.length).toBe(0); ??
    });
    

    // test('should render arrow toggle button', () => {
    //     const mockCallBack = jest.fn();
    //     const toggle = shallow(
    //         <DataContextProvider>
    //             <ToggleArrowsLink onClick={mockCallBack}>
    //                 Arrows
    //             </ToggleArrowsLink>
    //         </DataContextProvider>
    //    );
    //     toggle.simulate('click');
    //     // expect(mockCallBack.mock.calls.length).toBe(0); ????
    // });

   
    // something from source to get id of and maybe pass in above
    describe("filter function to work stuff out", () => {
        test("it should filter by a term (title)", () => {
          const shows = [
            { id: "a1", title: "Gaycation", episodes: 24, product_image_url:"/images/gaycation.jpg" },
            { id: "b2", title: "Huang's World", episodes: 12, product_image_url:"/images/huangsworld.jpg" },
            { id: "c3", title: "King of The Road", episodes: 8, product_image_url:"/images/kingoftheroad.jpg" }
          ];
    
          const output = [{ id: "a1", title: "Gaycation", episodes: 24, product_image_url:"/images/gaycation.jpg" },];
          expect(filterShow(shows, "Gaycation")).toEqual(output);
          expect(filterShow(shows, "GAYCATION")).toEqual(output);
      
        });
    });

    function filterShow(inputArr, filterShow) {
        const regex = new RegExp(filterShow, "id");
        return inputArr.filter(function(arrayElement) {
          return arrayElement.title.match(regex);
        });
    }
      
});


