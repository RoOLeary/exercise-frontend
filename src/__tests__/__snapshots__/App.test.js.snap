// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bczRLJ cGrDtW mainLayout"
>
  <div
    className="sc-iqcoie gsDiQy"
  >
    <ul
      className="sc-crXcEl kTeTAh"
    >
      <h1
        className="sc-evZas geJLAy"
      >
        Loading Shows...
      </h1>
    </ul>
    <hr
      className="sc-ksZaOG dxRPeS"
    />
    <div
      className="sc-breuTD ggXwcJ"
    >
      <a
        className="sc-ftvSup iSZRKp"
        onClick={[Function]}
      >
        Reset
      </a>
      <a
        className="sc-kDDrLX bqZRJm"
        onClick={[Function]}
      >
        Arrows
      </a>
    </div>
  </div>
</div>
`;
